[net] Fix possible accept bug

ghaerr opened the pull request:
@Mellvik,

I found a possible bug with listen/accept sockets. I haven't been able to reproduce any bug yet, but this code both fixes that potential issue, displays the cases where the socket is changed from accept/listen (should never be zero!) and also ignores packets in established state with a null socket (with error display).

Lets see what your code does with this, thank you!

ghaerr committed:
[net] Fix possible accept bug

ghaerr committed:
Add tons of status messages

ghaerr commented:
Hello @Mellvik,

This is pretty perplexing, I've reviewed tons of code and can't quite see what the problem is. Provided that it is not ftpd corrupting data, I think it may have something to do with multiple accepts happening almost simultaneously. That is, the accept that is happening in passive mode, as well as sometimes the parent process executing an accept (and waiting) for the next client, either before or after the passive accept(s). There is also a possibility of socket number reuse, as sockets are actually internal inodes and do get reused once closed. I'm about as far as I can imagine without being able to duplicate anything over here.

Can you describe your test in more detail? I have been reviewing ftpd, and don't know whether you are using multiple clients, or just multiple passive mode transfers (=accepts, right?) or both. This may be important if the problem ends up being the multiple accept issue described above, so I can guess more particularly what system calls might be happening.

The latest push adds tons of debug code, and should show when we get a zero socket. Please run your test and post what you find, should it show a zero or NULL socket message. Because of all the output, it could easily change the timing. 

We also need to make sure it is not ftpd overwriting some data, so changing around those buffer sizes might also be a good idea, to see whether that changes anything. (there is also a hardcoded 256 buffer size on one line FYI which should probably be changed to CMDBUFSIZ). Ensure that the -maout-stack=8192 option is set too, thanks.

Mellvik commented:
Perplexing is just what it is, @ghaerr - but maybe - your added debug prints is just what it takes.

First - you requested details about the test. Here we go:
- Idle system
- ftpd running - some times in debug mode, other times not.
- Connection from (interactive) client, logging in, client set to use passive mode
- By now the command channel has been active for a few commands, LOGIN, PASSVD, possibly TYPE (bin or ascii).
- The command channels is always active, stays that way for the duration of the session.
- No other commands from the client, just `ftp> dir /bin`, which is supposed to list the server's /bin directory - output from `ls -l`
- Which fails with the following output (before the lastest debug stuff, this is w/o ^P output, just `ftpd` debug output):
```
# ./ftpd -d
New Client Detected...
Child running - cmd chan is 5.
cmd: 'SYST' 29
cmd: 'PASV' 9
getsockname: adr 10.0.2.15, port 49821
227 Entering Passive Mode (10,0,2,15,194,157)
Accepted connection from 10.0.2.2/54122 on fd 8
cmd: 'LIST /bin' 26
LIST/NDIR: '/bin/ls -l /bin' fd 6
tcpdev_write: write to unknown socket
LIST: fd 6 len60
LIST: Data socket write error: Broken pipe 
```

- This is entirely repeatable, provided ELKS was just restarted. If the `ftpd`-child si exited and a new started, the sequence above will most likely succeed.
- With the new debug stuff in, the command sequence above succeeds. However, the problem manifests itself a little later, after transferring a few files (I'm including the entire listing, from boot via the successful commands to the final error, see comments in the listing, `ftpd` is running w/o debug in this case, just a few messages from `ftpd`):
- There may just be enough here for you to drill down, @ghaerr!

```
Running /etc/rc.d/rc.sys script
Starting networking on eth
ktcp -b 10.0.2.15 10.0.2.2 255.255.255.0
ktcp: ip 10.0.2.15, gateway 10.0.2.2, netmask 255.255.255.0
ktcp: ethernet 00.80.29.ef.d9.51 mtu 1500
Starting daemons telnetd sock_alloc 9a9e
NET(13) sys_listen sock 9a9e
tcp listen: port 23 sock[009a9e]
sock_alloc 9aee
INET(14) accept wait sock 9a9e -> newsock 9aee
httpd tcp accept: waiting on SYN, saving newsock[009aee]
sock_alloc 9b8e
NET(16) sys_listen sock 9b8e
tcp listen: port 80 sock[009b8e]
sock_alloc 9c2e
INET(17) accept wait sock 9b8e -> newsock 9c2e
ftpd tcp accept: waiting on SYN, saving newsock[009c2e]
sock_alloc 9cce
NET(19) sys_listen sock 9cce
tcp listen: port 21 sock[009cce]
sock_alloc 9d1e
INET(20) accept wait sock 9cce -> newsock 9d1e

tcp accept: waiting on SYN, saving newsock[009d1e]
ELKS built from commit ecd4d1c4
Mon Nov 29 12:45:14 2021


ELKS 0.4.0

login: tcp listen: got SYN clone sock[009cce]
tcpdev checkaccept: searching accept->newsock[009cce]
tcpdev checkaccept: found listen socket[009cce] newsocket[009d1e]
tcpdev checkaccept: accept SYN received before accept sock[009d1e]
INET(20) accepted sock 9cce -> newsock 9d1e
sock_alloc 9fee
INET(20) accept wait sock 9cce -> newsock 9fee
tcp accept: waiting on SYN, saving newsock[009fee]

login: sock_alloc a08e                  <-------------- Client connects
NET(27) sys_listen sock a08e
tcp listen: port 49821 sock[00a08e]
sock_alloc a0de
INET(27) accept wait sock a08e -> newsock a0de
tcp accept: waiting on SYN, saving newsock[00a0de]
tcp listen: got SYN clone sock[00a08e]
tcpdev checkaccept: searching accept->newsock[00a08e]
tcpdev checkaccept: found listen socket[00a08e] newsocket[00a0de]
tcpdev checkaccept: accept SYN received before accept sock[00a0de]
INET(27) accepted sock a08e -> newsock a0de
sock_free a08e
Accepted connection from 10.0.2.2/54226 on fd 5          <------ [SUCCESS] DIR command entered
sock_free a0de
sock_alloc a03e
NET(27) sys_listen sock a03e
tcp listen: port 49821 sock[00a03e]
sock_alloc a12e
INET(27) accept wait sock a03e -> newsock a12e
tcp accept: waiting on SYN, saving newsock[00a12e]
tcp listen: got SYN clone sock[00a03e]
tcpdev checkaccept: searching accept->newsock[00a03e]
tcpdev checkaccept: found listen socket[00a03e] newsocket[00a12e]
tcpdev checkaccept: accept SYN received before accept sock[00a12e]
INET(27) accepted sock a03e -> newsock a12e
sock_free a03e
Accepted connection from 10.0.2.2/54228 on fd 5     <---- [SUCCESS] Entering another DIR command after having changed the CWD (which is done via the command channel only).
sock_free a12e
sock_alloc b07e
NET(27) sys_listen sock b07e
tcp listen: port 49821 sock[00b07e]
sock_alloc b0ce
INET(27) accept wait sock b07e -> newsock b0ce
tcp listen: got SYN clone sock[00b07e]
tcpdev accept: accept SYN received before accept sock[00b0ce], ZEROING newsock
INET(27) accepted sock b07e -> newsock b0ce
sock_free b07e
Accepted connection from 10.0.2.2/54230 on fd 5  <------ [SUCCESS] Starting a send file into ftpd: 'put vi' - about 69kB
tcp: dropping packet, bad seqno: need 64241 got 62781 size 1460 <---- server delay when saving data block to floppy
tcp: dropping packet, bad seqno: need 64241 got 62781 size 1460 <---- causes retransmits from the client, this is normal
tcp: dropping packet, bad seqno: need 64241 got 62781 size 1460
tcp: dropping packet, bad seqno: need 64241 got 62781 size 1460
tcp: dropping packet, bad seqno: need 64241 got 62781 size 1460
tcp: application read too small after FIN, data_avail 1784
tcp: application read too small after FIN, data_avail 292
sock_free b0ce

login: root
# cd /tmp
# sum vi
10604    135               <----- just checking that the file transfer was OK
# sock_alloc 94fe
NET(27) sys_listen sock 94fe
tcp listen: port 49821 sock[0094fe]
sock_alloc 97ce
INET(27) accept wait sock 94fe -> newsock 97ce
tcp accept: waiting on SYN, saving newsock[0097ce]
tcp listen: got SYN clone sock[0094fe]
tcpdev checkaccept: searching accept->newsock[0094fe]
tcpdev checkaccept: found listen socket[0094fe] newsocket[0097ce]
tcpdev checkaccept: accept SYN received before accept sock[0097ce]
INET(27) accepted sock 94fe -> newsock 97ce
sock_free 94fe
Accepted connection from 10.0.2.2/54232 on fd 5  <--- [SUCCESS] new file transfer, request the same 'vi' file back from the server
sock_free 97ce
sock_alloc 981e            <---- got it OK, integrity checked again with 'sum'
NET(27) sys_listen sock 981e
tcp listen: port 49821 sock[00981e]
sock_alloc 990e
INET(27) accept wait sock 981e -> newsock 990e
tcp listen: got SYN clone sock[00981e]
tcpdev accept: accept SYN received before accept sock[00990e], ZEROING newsock
INET(27) accepted sock 981e -> newsock 990e
sock_free 981e
Accepted connection from 10.0.2.2/54234 on fd 5   <----- [SUCCESS] new transfer, this time the command is 'mput xd?' which would send about 10 files to the server. 
<----- NOTE: The server or the FTP protocol has no concept of multi-file transfer, it's all done by the client sending repeated GET or PUT commands.
tcp: application read too small after FIN, data_avail 508
sock_free 990e
sock_alloc 9d6e      <---- file transfer completed successfully, client immediately requests another transfer
<----- NOTE: The back-to-back nature of this sequence does not seemingly influence of trigger the error.
<----- REMEBER: The 'normal' error situation, which did not happen this time, is just a DIR command,
<----- No activity in front, no activity shortly after.
NET(27) sys_listen sock 9d6e
tcp listen: port 49821 sock[009d6e]
sock_alloc 959e
INET(27) accept wait sock 9d6e -> newsock 959e
tcp listen: got SYN clone sock[009d6e]
tcpdev checkaccept: searching accept->newsock[009d6e] 
tcpdev checkaccept: found listen socket[009d6e] newsocket[000000] <---- [FAILURE] apparently this is about where the error happens.
tcpdev checkaccept: accept SYN received before accept sock[000000]
INET(27) accepted sock 9d6e -> newsock 959e
tcp established: NULL SOCKET
sock_free 9d6e
Accepted connection from 10.0.2.2/54236 on fd 5 <---- this is where the accept() call returns too ftpd, with a valid file descriptor
tcp accept: waiting on SYN, saving newsock[00959e]
tcp established: NULL SOCKET
tcp: dropping packet, bad seqno: need 1 got 1461 size 540
tcp: dropping packet, bad seqno: need 1 got 2001 size 0
tcp established: NULL SOCKET
tcp established: NULL SOCKET
tcp established: NULL SOCKET
tcp established: NULL SOCKET
tcp established: NULL SOCKET
tcp established: NULL SOCKET
tcp established: NULL SOCKET
tcp established: NULL SOCKET
```

ghaerr commented:
Hello @Mellvik,

Thanks for the fantastic trace. I'm pretty sure I see the problem - it occurs when the network receives a SYN from the remote (starting a new connection) *before* ftpd has had a chance to execute the `accept` system call. The accept system calls passes the new socket to ktcp, and that hasn't happened yet, so it's null. Basically, the ELKS applications are running asynchronously from the network, and the accept needs to happen in sync.

I'm pushing a fix shortly that hopefully fixes the problem; it delays the accept response until it actually has the new socket.

I'm not exactly sure if this is the problem for the broken pipe error above, we'll see.

Can you post a .zip version of your latest ftpd.c you're testing with? Thank you!

ghaerr committed:
Delay accept until have new socket

Mellvik commented:
Well, that is great news, @ghaerr!

There is one thing I forgot to mention:
With the ktcp/kernel changes that made the tracing possible, the following strange thing happens:
- ELKS boots normally, starts the daemons and elicits the new tracing-messages.
- I want to enable debugging in ftpd, so I kill the default ftpd and restart a new one from the command line.
- The new one loops in fork, as if it's getting new connections all the time, until elks runs out of process slots. It actually continues for a lot longer, getting errors about 0 slots available

What I found particularly interesting about this is not the looping, but the fact that it doesn't happen the first time the daemon is startet (net start), only the 2nd time.

Anyway, this may be gone in your latest update.

And one more thing I discovered while working on the new ftp client for elks:
From the model I used when creating the PASV support in ftpd, I inherited something that may be a bug: I'm (re)using the same sockaddr structure all the way through - socket(), bind(), listen(), accept(). Since accept() is indeed a new socket, I assume it doesn't need anything other than the file descriptor, and reusing the sockaddr structure may be harmless - or a bug.

Another inheritance is the second argument to the listen() call. I've used 1 in PASV, 3 in main(), and found other programs using 1024 or (BSD default) 128. 
Does this argument matter in ELKS?

I'm about to test the changes right now.
Zip file coming, and changes to qemu.sh to create a test environment.


—M


> 29. nov. 2021 kl. 16:29 skrev Gregory Haerr ***@***.***>:
> 
> 
> Hello @Mellvik <https://github.com/Mellvik>,
> 
> Thanks for the fantastic trace. I'm pretty sure I see the problem - it occurs when the network receives a SYN from the remote (starting a new connection) before ftpd has had a chance to execute the accept system call. The accept system calls passes the new socket to ktcp, and that hasn't happened yet, so it's null. Basically, the ELKS applications are running asynchronously from the network, and the accept needs to happen in sync.
> 
> I'm pushing a fix shortly that hopefully fixes the problem; it delays the accept response until it actually has the new socket.
> 
> I'm not exactly sure if this is the problem for the broken pipe error above, we'll see.
> 
> Can you post a .zip version of your latest ftpd.c you're testing with? Thank you!
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub <https://github.com/jbruchon/elks/pull/1035#issuecomment-981743895>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/AA3WGOBMTXBKNEVKTOIFWFDUOOL4VANCNFSM5I5W35JA>.
> Triage notifications on the go with GitHub Mobile for iOS <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675> or Android <https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>. 
>

ghaerr commented:
> The new one loops in fork, as if it's getting new connections all the time, until elks runs out of process slots. It actually continues for a lot longer, getting errors about 0 slots available

> What I found particularly interesting about this is not the looping, but the fact that it doesn't happen the first time the daemon is startet (net start), only the 2nd time.

That may be a new bug I just introduced, it won't be fixed in the latest update. Lets hold on this for a minute. I will try to repeat once I get your ftpd.c. This is related to accept, it seems that ftpd is not hanging in the outer-loop accept. Not sure why yet.


> I inherited something that may be a bug: I'm (re)using the same sockaddr structure all the way through - socket(), bind(), listen(), accept(). Since accept() is indeed a new socket, I assume it doesn't need anything other than the file descriptor, and reusing the sockaddr structure may be harmless - or a bug.

Reusing the sockaddr struct should be no problem - they are never referenced after the system call returns.


> Another inheritance is the second argument to the listen() call. I've used 1 in PASV, 3 in main(), and found other programs using 1024 or (BSD default) 128.
> Does this argument matter in ELKS?

It doesn't matter. It is supposed to be the number of SYNs coming in (new connections to listen) that will be SYN'd, that is accepted. Using 1 is probably the way to go with ELKS for the time being, since I am now deep in ktcp looking at how this is handled, and ktcp will "clone" any number of control blocks, one for each incoming SYN if received on a listen socket. This will definitely run ktcp out of memory. Lets not worry about all that right now.

> I'm about to test the changes right now.

Thanks for the quick feedback! I'm very interested to know whether that fixes the accept issue, then we can deal with the new ftpd loop issue and/or other issues.

Mellvik commented:
Well, how's a monday for good news, @ghaerr.

For the first time - MGETs and MPUTs work on physical HW. Extensive testing still pending - but this looks VERY promising. And - not that it matters - the first tests were run on two machines in parallell. Turns out to improve turnaround time, test on one while the other reboots (30 secs). :-)

And thanks for the update on my questions - I'll work that into the source.


More testing pending. Source coming.

BTW - any theory why this was working on QEMU before the fix?


—Mellvik


> 29. nov. 2021 kl. 17:52 skrev Gregory Haerr ***@***.***>:
> 
> 
> The new one loops in fork, as if it's getting new connections all the time, until elks runs out of process slots. It actually continues for a lot longer, getting errors about 0 slots available
> 
> What I found particularly interesting about this is not the looping, but the fact that it doesn't happen the first time the daemon is startet (net start), only the 2nd time.
> 
> That may be a new bug I just introduced, it won't be fixed in the latest update. Lets hold on this for a minute. I will try to repeat once I get your ftpd.c. This is related to accept, it seems that ftpd is not hanging in the outer-loop accept. Not sure why yet.
> 
> I inherited something that may be a bug: I'm (re)using the same sockaddr structure all the way through - socket(), bind(), listen(), accept(). Since accept() is indeed a new socket, I assume it doesn't need anything other than the file descriptor, and reusing the sockaddr structure may be harmless - or a bug.
> 
> Reusing the sockaddr struct should be no problem - they are never referenced after the system call returns.
> 
> Another inheritance is the second argument to the listen() call. I've used 1 in PASV, 3 in main(), and found other programs using 1024 or (BSD default) 128.
> Does this argument matter in ELKS?
> 
> It doesn't matter. It is supposed to be the number of SYNs coming in (new connections to listen) that will be SYN'd, that is accepted. Using 1 is probably the way to go with ELKS for the time being, since I am now deep in ktcp looking at how this is handled, and ktcp will "clone" any number of control blocks, one for each incoming SYN if received on a listen socket. This will definitely run ktcp out of memory. Lets not worry about all that right now.
> 
> I'm about to test the changes right now.
> 
> Thanks for the quick feedback! I'm very interested to know whether that fixes the accept issue, then we can deal with the new ftpd loop issue and/or other issues.
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub <https://github.com/jbruchon/elks/pull/1035#issuecomment-981820146>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/AA3WGODIAUZVT3TDDMNTP5TUOOVU3ANCNFSM5I5W35JA>.
> Triage notifications on the go with GitHub Mobile for iOS <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675> or Android <https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>. 
>

Mellvik commented:
[ftpd.c.gz](https://github.com/jbruchon/elks/files/7619836/ftpd.c.gz)

```

 # Simultaneous telnet and http forwarding
-FWD="hostfwd=tcp:127.0.0.1:8080-10.0.2.15:80,hostfwd=tcp:127.0.0.1:2323-10.0.2.15:23"
+FWD="hostfwd=tcp:127.0.0.1:8021-10.0.2.15:21,hostfwd=tcp:127.0.0.1:2323-10.0.2.15:23,\
+hostfwd=tcp:127.0.0.1:8020-10.0.2.15:20,hostfwd=tcp:127.0.0.1:8041-10.0.2.15:49821,\
+hostfwd=tcp:127.0.0.1:8042-10.0.2.15:49822,hostfwd=tcp:127.0.0.1:8043-10.0.2.15:49823,\
+hostfwd=tcp:127.0.0.1:8044-10.0.2.15:49824,hostfwd=tcp:127.0.0.1:8045-10.0.2.15:49825"
+#FWD="hostfwd=tcp:127.0.0.1:8080-10.0.2.15:80,hostfwd=tcp:127.0.0.1:2323-10.0.2.15:23"
 # new style
 #NET="-net nic,model=ne2k_isa -net user,$FWD"
```

ghaerr commented:
Great news @Mellvik!

The fix I just posted requires that ktcp until it receives a socket from the ftpd accept call before continuing. That wasn't always happening. That is, the SYN on the network may have been received and ktcp processed it before ftpd actually executed accept (which passes the new socket to ktcp).

> any theory why this was working on QEMU before the fix?

It seems that because of the actual ELKS kernel scheduling running on different hardware, that the SYN was received and processed by ktcp before ftpd executed the accept statement. Its all a matter of timing and particular kernel scheduling as to which tasks get to run next. Remember that the TCP input processing by ktcp is asynchronous to the kernel scheduling the tasks which are requesting socket activity.

> What I found particularly interesting about this is not the looping, but the fact that it doesn't happen the first time the daemon is startet (net start), only the 2nd time.


If you can get a trace of the ftpd looping after running the first verses the second time, that would be great. What's happening for sure is that the outer loop accept in ftpd is not blocking for some reason, which then executes a fork etc etc and runs out of tasks.

Mellvik commented:
@ghearr,

trace included below. Not as long as the previous one, but cleaner.

As you can see, the sequence is
login
kill 20
ftpd -d


> 29. nov. 2021 kl. 18:15 skrev Gregory Haerr ***@***.***>:
> 
> 
> Great news @Mellvik <https://github.com/Mellvik>!
> 
> The fix I just posted requires that ktcp until it receives a socket from the ftpd accept call before continuing. That wasn't always happening. That is, the SYN on the network may have been received and ktcp processed it before ftpd actually executed accept (which passes the new socket to ktcp).
> 
> any theory why this was working on QEMU before the fix?
> 
> It seems that because of the actual ELKS kernel scheduling running on different hardware, that the SYN was received and processed by ktcp before ftpd executed the accept statement. Its all a matter of timing and particular kernel scheduling as to which tasks get to run next. Remember that the TCP input processing by ktcp is asynchronous to the kernel scheduling the tasks which are requesting socket activity.
> 
> What I found particularly interesting about this is not the looping, but the fact that it doesn't happen the first time the daemon is startet (net start), only the 2nd time.
> 
> If you can get a trace of the ftpd looping after running the first verses the second time, that would be great. What's happening for sure is that the outer loop accept in ftpd is not blocking for some reason, which then executes a fork etc etc and runs out of tasks.
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub <https://github.com/jbruchon/elks/pull/1035#issuecomment-981841890>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/AA3WGOH5446ZK5JR6MQQ6E3UOOYMLANCNFSM5I5W35JA>.
> Triage notifications on the go with GitHub Mobile for iOS <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675> or Android <https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>. 
> 

Starting networking on eth
ktcp -b 10.0.2.15 10.0.2.2 255.255.255.0
ktcp: ip 10.0.2.15, gateway 10.0.2.2, netmask 255.255.255.0
ktcp: ethernet 00.80.29.ef.d9.51 mtu 1500
Starting daemons telnetd sock_alloc 9a9e
NET(13) sys_listen sock 9a9e
tcp listen: port 23 sock[009a9e]
sock_alloc 9aee
INET(14) accept wait sock 9a9e -> newsock 9aee
httpd sock_alloc 9b8e
tcp accept: waiting on SYN, saving newsock[009aee]
NET(16) sys_listen sock 9b8e
tcp listen: port 80 sock[009b8e]
sock_alloc 9c2e
INET(17) accept wait sock 9b8e -> newsock 9c2e
ftpd sock_alloc 9cce
tcp accept: waiting on SYN, saving newsock[009c2e]
NET(19) sys_listen sock 9cce
tcp listen: port 21 sock[009cce]
sock_alloc 9d1e
INET(20) accept wait sock 9cce -> newsock 9d1e

tcp accept: waiting on SYN, saving newsock[009d1e]
ELKS built from commit ecd4d1c4
Mon Nov 29 18:24:40 2021


ELKS 0.4.0

login: root
# kill 20
sock_free 9d1e
# sock_free 9cce
ftpd -d
sock_alloc a17e
NET(28) sys_listen sock a17e
sock_alloc a1ce
INET(28) accept wait sock a17e -> newsock a1ce
INET(28) accepted sock a17e -> newsock a1ce
New Client Detected...
tcp listen: port 21 sock[00a17e]
sock_alloc a21e
INET(28) accept wait sock a17e -> newsock a21e
Child running - cmd chan is 5.
INET(28) accepted sock a17e -> newsock a21e
New Client Detected...
tcp accept: waiting on SYN, saving newsock[00a1ce]
sock_alloc a26e
INET(28) accept wait sock a17e -> newsock a26e
Child running - cmd chan is 5.
INET(28) accepted sock a17e -> newsock a26e
New Client Detected...
tcp accept: waiting on SYN, saving newsock[00a21e]
sock_alloc a2be
INET(28) accept wait sock a17e -> newsock a2be
Child running - cmd chan is 5.
INET(28) accepted sock a17e -> newsock a2be
New Client Detected...
sock_alloc a30e
INET(28) accept wait sock a17e -> newsock a30e
Child running - cmd chan is 5.
tcp accept: waiting on SYN, saving newsock[00a26e]
INET(28) accepted sock a17e -> newsock a30e
New Client Detected...
tcp accept: waiting on SYN, saving newsock[00a2be]
sock_alloc a35e
INET(28) accept wait sock a17e -> newsock a35e
Child running - cmd chan is 5.
INET(28) accepted sock a17e -> newsock a35e
New Client Detected...
tcp accept: waiting on SYN, saving newsock[00a30e]
sock_alloc a3ae
INET(28) accept wait sock a17e -> newsock a3ae
Child running - cmd chan is 5.
INET(28) accepted sock a17e -> newsock a3ae
New Client Detected...
tcp accept: waiting on SYN, saving newsock[00a35e]
Only 1 task slots
sock_alloc a3fe
INET(28) accept wait sock a17e -> newsock a3fe
Child running - cmd chan is 5.
INET(28) accepted sock a17e -> newsock a3fe
New Client Detected...
tcp accept: waiting on SYN, saving newsock[00a3ae]
Only 0 task slots
sock_free a3fe
sock_alloc a44e
INET(28) accept wait sock a17e -> newsock a44e
tcp accept: waiting on SYN, saving newsock[00a3fe]
INET(28) accepted sock a17e -> newsock a44e
New Client Detected...
Only 0 task slots
sock_free a44e
sock_alloc a49e
INET(28) accept wait sock a17e -> newsock a49e
tcp accept: waiting on SYN, saving newsock[00a44e]
INET(28) accepted sock a17e -> newsock a49e
New Client Detected...
Only 0 task slots
sock_free a49e
sock_alloc a4ee
INET(28) accept wait sock a17e -> newsock a4ee
tcp accept: waiting on SYN, saving newsock[00a49e]
INET(28) accepted sock a17e -> newsock a4ee
New Client Detected...
Only 0 task slots
sock_free a4ee
sock_alloc a53e
INET(28) accept wait sock a17e -> newsock a53e
tcp accept: waiting on SYN, saving newsock[00a4ee]
INET(28) accepted sock a17e -> newsock a53e
New Client Detected...
Only 0 task slots
sock_free a53e
sock_alloc a58e
INET(28) accept wait sock a17e -> newsock a58e
tcp accept: waiting on SYN, saving newsock[00a53e]
INET(28) accepted sock a17e -> newsock a58e
New Client Detected...
tcp accept: waiting on SYN, saving newsock[00a58e]
Only 0 task slots
sock_free a58e
sock_alloc a5de
INET(28) accept wait sock a17e -> newsock a5de
tcpdev_write: write to unknown socket
INET(28) accepted sock a17e -> newsock a5de
sock_free a5de
Accept error:: tcp accept: waiting on SYN, saving newsock[00a5de]
tcpdev_write: write to unknown socket
Broken pipe 
sock_free a17e
# tcpdev_write: write to unknown socket
tcpdev_write: write to unknown socket
tcpdev_write: write to unknown socket
tcpdev_write: write to unknown socket
tcpdev_write: write to unknown socket

ghaerr commented:
> As you can see, the sequence is
login
kill 20
ftpd -d

When you `kill` ftpd, I can't see what's happening in the close. Can you do one more trace with ^P on at login? That will show what is being released from the previous ftpd. Also, make sure (with ps) that the previous ftpd did in fact die.

Later today I will push a prettier version of the fixes that will be using DEBUG_ACCEPT (default on only with ^P) so we can test without the printf statements.

thanks!

ghaerr commented:
> When you kill ftpd, I can't see what's happening in the close. Can you do one more trace with ^P on at login? That will show what is being released from the previous ftpd.

No need to send another trace, I have the ftpd loop bug duplicating over here. Thanks for all the testing @Mellvik, we're making big progress!

Mellvik commented:
Happy to hear that, @ghaerr. However small, this feels like milestone.

I plan to push a 'mergeable' ftpd tomorrow. A ftp client is in the works, will be very helpfull for the elks back-to-back testing.

Thank you for taking the deep dive into this rather elusive problem. 


-Mellvik

> 29. nov. 2021 kl. 19:59 skrev Gregory Haerr ***@***.***>:
> 
> ﻿
> When you kill ftpd, I can't see what's happening in the close. Can you do one more trace with ^P on at login? That will show what is being released from the previous ftpd.
> 
> No need to send another trace, I have the ftpd loop bug duplicating over here. Thanks for all the testing @Mellvik, we're making big progress!
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or unsubscribe.
> Triage notifications on the go with GitHub Mobile for iOS or Android.

ghaerr committed:
Cleanup debug messages, add listen semaphore

ghaerr commented:
@Mellvik,

I cleaned up all the debug messages so they will only display with ^P. This is almost exactly the same as what you tested yesterday, except I added a semaphore around `listen` for some additional multitasking protection, which likely will have no effect.

> I plan to push a 'mergeable' ftpd tomorrow.

Nice! Please test it against this PR with the last commit included. If it works well, then lets commit both this PR and your new ftpd PR. I would suggest opening a completely new PR with your working version, and closing #1030.

> A ftp client is in the works, will be very helpfull for the elks back-to-back testing.

Can't wait to see it. Will it use active or passive mode? When we start testing ELKS-to-ELKS, we should also put ftpget/put in the mix.

>  I kill the default ftpd and restart a new one from the command line.
> The new one loops in fork...

This new "restart" problem is definitely not fixed in this PR. I started looking into it and got nowhere. I don't want to disturb what we've hopefully gotten to work well. If we get your ftpd and this PR finalized soon, then I'll open a separate issue for the daemon restart problem.

Mellvik commented:
More good news, @ghaerr - 
> I cleaned up all the debug messages so they will only display with ^P. This is almost exactly the same as what you tested yesterday, except I added a semaphore around listen for some additional multitasking protection, which likely will have no effect.
> 
> I plan to push a 'mergeable' ftpd tomorrow.
> 
> Nice! Please test it against this PR with the last commit included. If it works well, then lets commit both this PR and your new ftpd PR. I would suggest opening a completely new PR with your working version, and closing #1030 <https://github.com/jbruchon/elks/pull/1030>.
> 
Testing is progressing smoothly. Passive mode is now just as stable as active mode on the 386. On the 286 I'm experiencing system hang during file transfers, which needs to be investigated, but I have no reason to believe it has anything to do with the accept problem, probably not networking at all. It may be FS errors on the FAT file system I'm transferring to. I've seen that before.

A couple more messages can go now, such as 
tcp: port 49821 already bound, rejecting (use SO_REUSEADDR?) 
tcp: application read too small after FIN, data_avail 508 
Possibly (although this one may be useful for the back-toi-back testing)
tcp: dropping packet, bad seqno: need 182501 got 181041 size 1460

I suggest we keep the 
tcp retransmit: 
message for now.
> A ftp client is in the works, will be very helpfull for the elks back-to-back testing.
> 
> Can't wait to see it. Will it use active or passive mode?
> 
The client will support both - and the basic commands - get/put/mget/mput/dir/cwd/pwd/del/mkdir and administrative commands like user/pass/type/
> When we start testing ELKS-to-ELKS, we should also put ftpget/put in the mix.
> 
Absolutely - the script(s) you have created would be very useful.
> I kill the default ftpd and restart a new one from the command line.
> The new one loops in fork...
> 
> This new "restart" problem is definitely not fixed in this PR. I started looking into it and got nowhere. I don't want to disturb what we've hopefully gotten to work well. If we get your ftpd and this PR finalized soon, then I'll open a separate issue for the daemon restart problem.
> 
Actually - this problem is now magically gone. So I guess we'll just leave it at that - for now.

—M

ghaerr commented:
>  Passive mode is now just as stable as active mode on the 386.

Fantastic news. What a whirlwind the last couple weeks have been. I'm glad it's all working well now.

> A couple more messages can go now

I'll remove them in a next cleanup commit.

> This new "restart" problem is definitely not fixed in this PR.
> this problem is now magically gone.

I broke it by adding a semaphore release which was in error, which was removed in my last push.

However, I still saw the problem with all the debugging on, and then, on final testing, kill and restart seemed to be working again (I tested with telnetd). So it's still likely there. I'll remove more of the commented out debug code I added in a cleanup PR, since it's not an obvious coding error. The best workaround for now is to completely restart networking using `net stop` and `net start`.
